crates/resolver/src/error.rs

```diff
@@ -128,7 +128,24 @@ impl ResolveError {
     /// A conversion to determine if the response is an error
     pub fn from_response(response: DnsResponse, trust_nx: bool) -> Result<DnsResponse, Self> {
         match response.response_code() {
-            response_code @ ResponseCode::ServFail | response_code @ ResponseCode::Refused => {
+            response_code @ ResponseCode::ServFail
+            | response_code @ ResponseCode::Refused 
+            | response_code @ ResponseCode::FormErr
+            | response_code @ ResponseCode::NotImp
+            | response_code @ ResponseCode::YXDomain
+            | response_code @ ResponseCode::YXRRSet
+            | response_code @ ResponseCode::NXRRSet
+            | response_code @ ResponseCode::NotAuth
+            | response_code @ ResponseCode::NotZone
+            | response_code @ ResponseCode::BADVERS
+            | response_code @ ResponseCode::BADSIG
+            | response_code @ ResponseCode::BADKEY
+            | response_code @ ResponseCode::BADTIME
+            | response_code @ ResponseCode::BADMODE
+            | response_code @ ResponseCode::BADNAME
+            | response_code @ ResponseCode::BADALG
+            | response_code @ ResponseCode::BADTRUNC
+            | response_code @ ResponseCode::BADCOOKIE => {
```

Comment:
Maybe this `match` should be in a `ResponseCode` method instead? Like `ResponseCode::should_retry()` or something. Feels like this is tightly coupled to the `ResponseCode` variants/definition.

Reply:
Yeah, I see what you mean. It is tightly coupled to the `ResponseCode` variants, but `ResponseCode` is in the proto crate and IIUC is really meant to map directly on to the RFCs (e.g. [1035](https://datatracker.ietf.org/doc/html/rfc1035#section-4.1.1) for the initial response codes), and I think it would be better to leave more "opinionated" stuff, like which response codes should elicit a retry, which is more subjective, in the resolver crate.

Thoughts?

Reply:
My thoughts are that I'd probably prioritize low coupling over exact alignment of the proto crate with what's defined in the RFC, and also that maybe this is a response to the name `should_retry()` in particular; maybe the name could be less opinionated?

But in the end, maybe @bluejekyll should weigh in on this.

Reply:
> maybe this is a response to the name should_retry() in particular

That's a good point, it could be something more like `retryable`.

> I'd probably prioritize low coupling over exact alignment of the proto crate with what's defined in the RFC

That's fair. I'm of the opinion that the semantics of this match have more to do with `ResolveError` than with `ResponseCode` — we special case `NxDomain | NoError` responses that contain no records, etc. I guess the logic that would make sense to be moved into `ResponseCode` is whether a response code is considered an error, but here we consider basically every response code except `NoError` an error that could lead to retry.

So maybe there should be something like this in `ResponseCode`:

```
impl ResponseCode {
    ...
    pub fn is_error(&self) -> bool {
        match self {
            NoError | Unknown(_) => false,
            FormErr | ServFail | NXDomain | NotImp | Refused | YXDomain | YXRRSet | NXRRSet
            | NotAuth | NotZone | BADVERS | BADSIG | BADKEY | BADTIME | BADMODE | BADNAME
            | BADALG | BADTRUNC | BADCOOKIE => true,
        }
    }
}
```

And then `ResolveError::from_response` can call this. What do you think?

Reply:
Fair enough, but if that doesn't materially improve your current change (to make more responses retryable) maybe it should be a different PR.

Reply:
SGTM, I can follow up with another PR to do this.